src/checker/checker.rs

```diff
@@ -0,0 +1,372 @@
+        // all decoded rows and return the index of the first matching row
+        let matching = decoded_rows.iter().position(|row| {
+            match row {
+                Ok(row) => {
+                    // Check that all columns specified in the check
+                    // have the expected value in the decoded row
+                    for col in &row_check.columns {
+                        let result = row.get(&col.column).unwrap();
+                        if result.to_string() != col.value {
+                            return false;
+                        }
+                    }
+                    // If all match, then return row index
+                    true
+                }
+                Err(e) => {
+                    error!("{}", e);
+                    return false;
+                }
+            }
+        });
+
+        // Remove item that we found (if any)
+        if let Some(matched) = matching {
+            decoded_rows.remove(matched).expect("Row must exist");
+        } else {
+            // No row matched our check, the validation should fail
+            return false;
+        }
+    }
+
+    // If there are any rows left in checked, the validation should fail
+    if decoded_rows.len() != 0 {
+        false
+    } else {
+        true
+    }
+}
+
+fn decode(row: &Row) -> Result<HashMap<String, String>, String> {
+    let mut out = HashMap::new();
+    for (i, col) in row.columns().iter().enumerate() {
+        let ty = col.type_();
+        let conversion = match *ty {
+            Type::BOOL => row.get::<_, Option<bool>>(i).map(|x| x.to_string()),
+            Type::CHAR | Type::TEXT => row.get::<_, Option<String>>(i),
+            Type::BYTEA => row.get::<_, Option<Vec<u8>>>(i).map(|x| {
+                let s = x.into_iter().map(ascii::escape_default).flatten().collect();
+                String::from_utf8(s).unwrap()
+            }),
+            Type::INT4 => row.get::<_, Option<i32>>(i).map(|x| x.to_string()),
+            Type::INT8 => row.get::<_, Option<i64>>(i).map(|x| x.to_string()),
+            Type::NUMERIC => row.get::<_, Option<Numeric>>(i).map(|x| x.to_string()),
+            Type::TIMESTAMP => row
+                .get::<_, Option<chrono::NaiveDateTime>>(i)
+                .map(|x| x.to_string()),
+            Type::TIMESTAMPTZ => row
+                .get::<_, Option<chrono::DateTime<chrono::Utc>>>(i)
+                .map(|x| x.to_string()),
+            Type::DATE => row
+                .get::<_, Option<chrono::NaiveDate>>(i)
+                .map(|x| x.to_string()),
+            Type::INTERVAL => row.get::<_, Option<Interval>>(i).map(|x| x.to_string()),
+            _ => return Err(format!("unable to handle SQL type: {:?}", ty)),
+        }
+        .unwrap_or_else(|| "<null>".into());
+        out.insert(String::from(col.name()), conversion);
+    }
+    Ok(out)
+}
+
+fn spawn_query_thread(mz_url: String, check: Check) -> Vec<JoinHandle<()>> {
```

Comment:
Maybe `spawn_query_threads` since you will get `check.thread_count` threads from this?